lib/libprocstat/libprocstat.c

```diff
@@ -399,6 +405,84 @@ procstat_getc18n(struct procstat *procstat, struct kinfo_proc *kp,
 	return (-1);
 }
 
+int
+procstat_getcompartments(struct procstat *procstat, struct kinfo_proc *kp,
+    struct cheri_c18n_compart **comparts, size_t *ncompartsp)
+{
+	int name[4];
+	char *inbuf;
+	size_t n, i, cur, size;
+	struct cheri_c18n_compart *outbuf, *cp;
+
+	if (comparts == NULL || ncompartsp == NULL)
+		goto out;
+
+	switch (procstat->type) {
+	case PROCSTAT_KVM:
+		warnx("kvm method is not supported");
+		goto out;
+
+	case PROCSTAT_SYSCTL:
+		break;
+
+	case PROCSTAT_CORE:
+		warnx("core method is not supported");
+		goto out;
+
+	default:
+		warnx("unknown access method: %d", procstat->type);
+		goto out;
+	}
+
+	name[0] = CTL_KERN;
+	name[1] = KERN_PROC;
+	name[2] = KERN_PROC_C18N_COMPARTS;
+	name[3] = kp->ki_pid;
+
+	/* Estimate the size of the input buffer. */
+	if (sysctl(name, nitems(name), NULL, &size, NULL, 0) != 0) {
+		if (errno != ESRCH && errno != EPERM && errno != ENOEXEC)
+			warn("sysctl(kern.proc.c18n_compartments)");
+		goto out;
+	}
+
+	/* Receive the input buffer and count the number of items. */
+	inbuf = malloc(size);
+	if (inbuf == NULL)
+		goto out;
+	if (sysctl(name, nitems(name), inbuf, &size, NULL, 0) != 0) {
+		if (errno != ESRCH && errno != EPERM && errno != ENOEXEC)
+			warn("sysctl(kern.proc.c18n_compartments)");
+		goto out_free;
+	}
+	for (n = 0, cur = 0; cur < size; ++n) {
+		cp = (struct cheri_c18n_compart *)(void *)&inbuf[cur];
+		cur += cp->ccc_structsize;
+	}
+
+	/* Unpack elements of the input buffer into the output buffer. */
+	outbuf = malloc(n * sizeof(*outbuf));
```

Comment:
My one suggestion here would be to use calloc() so that the extra padding between structures is zeroed.  However, that can be fixed up as a later commit after merging.